proxy/server/server.go

```diff
@@ -270,30 +282,86 @@ func (s *T) newTLSCfg(hostCfgs map[engine.HostKey]engine.Host) (*tls.Config, err
 
 	defaultHostName := ""
 	pairs := map[string]tls.Certificate{}
+	getCertFuncs := map[string]getCertificateFunc{}
+
 	for _, hostCfg := range hostCfgs {
 		if hostCfg.Settings.Default {
 			defaultHostName = hostCfg.Name
 		}
-		c := hostCfg.Settings.KeyPair
-		if c == nil {
-			continue
-		}
-		keyPair, err := tls.X509KeyPair(c.Cert, c.Key)
-		if err != nil {
-			return nil, err
-		}
-		if hostCfg.Settings.OCSP.Enabled {
-			log.Infof("%v OCSP is enabled for %v, resolvers: %v", s, hostCfg, hostCfg.Settings.OCSP.Responders)
-			r, err := s.stapler.StapleHost(&hostCfg)
+		if hostCfg.Settings.KeyPair != nil {
+			c := hostCfg.Settings.KeyPair
+			if hostCfg.Settings.AutoCert != nil {
+				log.Warningf("Host %v has a KeyPair and AutoCert enabled. KeyPair takes precedence. Autocert generation disabled.", hostCfg.Name)
+			}
+			keyPair, err := tls.X509KeyPair(c.Cert, c.Key)
 			if err != nil {
-				log.Warningf("%v failed to staple %v, error %v", s, hostCfg, err)
-			} else if r.Response.Status == ocsp.Good || r.Response.Status == ocsp.Revoked {
-				keyPair.OCSPStaple = r.Staple
-			} else {
-				log.Warningf("%s got undefined status from OCSP responder: %v", s, r.Response.Status)
+				return nil, err
+			}
+			if hostCfg.Settings.OCSP.Enabled {
+				log.Infof("%v OCSP is enabled for %v, resolvers: %v", s, hostCfg, hostCfg.Settings.OCSP.Responders)
+				r, err := s.stapler.StapleHost(&hostCfg)
+				if err != nil {
+					log.Warningf("%v failed to staple %v, error %v", s, hostCfg, err)
+				} else if r.Response.Status == ocsp.Good || r.Response.Status == ocsp.Revoked {
+					keyPair.OCSPStaple = r.Staple
+				} else {
+					log.Warningf("%s got undefined status from OCSP responder: %v", s, r.Response.Status)
+				}
+			}
+			pairs[hostCfg.Name] = keyPair
+		} else if hostCfg.Settings.AutoCert != nil {
+			ac := hostCfg.Settings.AutoCert
+			if s.autoCertMgrs == nil {
+				s.autoCertMgrs = make([]*autocert.Manager, 0, 0)
+			}
+
+			// Each host gets its own Autocert Manager - this allows individual
+			// certs to use different autocert authorities, as well as auth keys
+			autoCertMgr := &autocert.Manager{
+				Prompt:     autocert.AcceptTOS,
+				Cache:      s.autoCertCache,
+				HostPolicy: autocert.HostWhitelist(hostCfg.Name),
+				Email:      ac.Email,
 			}
+
+			if ac.RenewBeforeSeconds > 0 {
+				autoCertMgr.RenewBefore = time.Duration(ac.RenewBeforeSeconds) * time.Second
+			}
+
+			if ac.Key != "" || ac.DirectoryURL != "" {
+				autoCertMgr.Client = &acme.Client{
```

Comment:
If `ac.DirectoryURL != ""` and `ac.Key == ""` then we create a client but do not generate a key, and autoCertMgr won't do that for us. It only generates a key if `autoCertMgr.Client == nil`. We need a test that exposes this but, and then fix it.

Reply:
Scratch that. After closely examining the AutoCert code I see that Client.Key is generate if it is empty in `autocert.(*autocert.Manager).acmeClient()`

Reply:
How about changing this function a bit:
```go
func (s *T) newTLSCfg(hostCfgs map[engine.HostKey]engine.Host) (*tls.Config, error) {
	config, err := s.lsnCfg.TLSConfig()
	if err != nil {
		return nil, err
	}

	if config.NextProtos == nil {
		// "h2" is required in order to enable HTTP 2: https://golang.org/src/net/http/server.go
		config.NextProtos = []string{"h2", "http/1.1"}
	}

	defaultHostName := ""
	pairs := map[string]tls.Certificate{}
	getCertFuncs := map[string]getCertificateFunc{}

	for _, hostCfg := range hostCfgs {
		if hostCfg.Settings.Default {
			defaultHostName = hostCfg.Name
		}
		if hostCfg.Settings.KeyPair != nil {
			c := hostCfg.Settings.KeyPair
			if hostCfg.Settings.AutoCert != nil {
				log.Warningf("Host %v has a KeyPair and AutoCert enabled. KeyPair takes precedence. Autocert generation disabled.", hostCfg.Name)
			}
			keyPair, err := tls.X509KeyPair(c.Cert, c.Key)
			if err != nil {
				return nil, err
			}
			if hostCfg.Settings.OCSP.Enabled {
				log.Infof("%v OCSP is enabled for %v, resolvers: %v", s, hostCfg, hostCfg.Settings.OCSP.Responders)
				r, err := s.stapler.StapleHost(&hostCfg)
				if err != nil {
					log.Warningf("%v failed to staple %v, error %v", s, hostCfg, err)
				} else if r.Response.Status == ocsp.Good || r.Response.Status == ocsp.Revoked {
					keyPair.OCSPStaple = r.Staple
				} else {
					log.Warningf("%s got undefined status from OCSP responder: %v", s, r.Response.Status)
				}
			}
			pairs[hostCfg.Name] = keyPair
		} else if hostCfg.Settings.AutoCert != nil {
			ac := hostCfg.Settings.AutoCert
			if s.autoCertMgrs == nil {
				s.autoCertMgrs = make([]*autocert.Manager, 0, 0)
			}

			// Each host gets its own Autocert Manager - this allows individual
			// certs to use different autocert authorities, as well as auth keys
			autoCertMgr := &autocert.Manager{
				Client: &acme.Client{
					// Let's Encrypt will be used if ac.DirectoryURL is "".
					DirectoryURL: ac.DirectoryURL,
				},
				Prompt:     autocert.AcceptTOS,
				Cache:      s.autoCertCache,
				HostPolicy: autocert.HostWhitelist(hostCfg.Name),
				Email:      ac.Email,
			}

			if ac.RenewBeforeSeconds > 0 {
				autoCertMgr.RenewBefore = time.Duration(ac.RenewBeforeSeconds) * time.Second
			}

			if ac.Key != "" {
				block, _ := pem.Decode([]byte(ac.Key))
				if block == nil {
					log.Errorf("Autocert Key PEM Block for Host %s is invalid", hostCfg.Name)
					continue
				}
				switch block.Type {
				case "RSA PRIVATE KEY":
					autoCertMgr.Client.Key, err = x509.ParsePKCS1PrivateKey(block.Bytes)
					if err != nil {
						log.Errorf("Error parsing Autocert Key block of type %s, for Host %s, as an RSA Private Key: %v", block.Type, hostCfg.Name, err)
						continue
					}
				case "EC PRIVATE KEY":
					autoCertMgr.Client.Key, err = x509.ParseECPrivateKey(block.Bytes)
					if err != nil {
						log.Errorf("Error parsing Autocert Key block of type %s, for Host %s, as an EC Private Key: %v", block.Type, hostCfg.Name, err)
						continue
					}
				default:
					log.Errorf("AutoCert Private Key for Host %s is of unrecognized type: %s. Supported types"+
						"are RSA PRIVATE KEY and EC PRIVATE KEY.", hostCfg.Name, block.Type)
					continue
				}
			}

			getCertFuncs[hostCfg.Name] = autoCertMgr.GetCertificate
		}
	}

	// Add static certificates to the TLS config.
	config.Certificates = make([]tls.Certificate, 0, len(pairs))
	if defaultHostName != "" {
		keyPair, ok := pairs[defaultHostName]
		if !ok {
			return nil, errors.Errorf("default host '%s' certificate is not passed", defaultHostName)
		}
		config.Certificates = append(config.Certificates, keyPair)
	}
	for hostName, keyPair := range pairs {
		if hostName != defaultHostName {
			config.Certificates = append(config.Certificates, keyPair)
		}
	}
	config.BuildNameToCertificate()

	// Ensure generation/renewal of certificates for AutoCert enabled hosts.
	config.GetCertificate = func(info *tls.ClientHelloInfo) (*tls.Certificate, error) {
		if getCertificateFunc, ok := getCertFuncs[info.ServerName]; ok {
			cert, err := getCertificateFunc(info)
			if err != nil {
				log.Errorf("Failed to generate Autocert for ServerName: %s. Error: %v.", info.ServerName, err)
			}
			return cert, err
		}
		return nil, nil
	}

	return config, nil
}
```